api/models/reviewDao.js

```diff
@@ -0,0 +1,157 @@
+const dbDataSource = require("./dataSource");
+
+const createReview = async (userId, rating, comment, movieId) => {
+  await dbDataSource.query(`set FOREIGN_KEY_CHECKS = 0`);
+  await dbDataSource.query(
+    `
+    INSERT INTO movie_reviews (
+      users_id, rating, comment, movies_id
+    ) VALUES (
+      ?, ?, ?, ?
+    )
+    `,
+    [userId, rating, comment, movieId]
+  );
+  await dbDataSource.query(`set FOREIGN_KEY_CHECKS = 1`);
+
+  let totalReview = await dbDataSource.query(
+    `
+    SELECT COUNT(id) as id
+    FROM movie_reviews
+    WHERE movies_id = ?
+    `,
+    [movieId]
+  );
+  totalReview = totalReview[0].id;
+
+  let averageRating = await dbDataSource.query(
+    `
+    SELECT SUM(rating) as average_rating
+    FROM movie_reviews
+    WHERE movies_id = ?
+    `,
+    [movieId]
+  );
+  averageRating = averageRating[0].average_rating;
+
+  await dbDataSource.query(
+    `
+    UPDATE movies
+    SET average_rating = ?
+    WHERE id = ?
+    `,
+    [averageRating / totalReview, movieId]
+  );
+};
+
+const getReview = async (movieId) => {
+  const result = await dbDataSource.query(
+    `
+    SELECT rating, comment, u.nickname
+    FROM movie_reviews as mr
+    JOIN users as u
+    ON mr.users_id = u.id
+    WHERE mr.movies_id = ?
+    `,
+    [movieId]
+  );
+  return result;
+};
+
+const updateReview = async (comment, checkReview) => {
+  const result = await dbDataSource.query(
+    `
+    UPDATE movie_reviews 
+    SET comment = ?
+    WHERE movie_reviews.id = ?
+    `,
+    [comment, checkReview]
+  );
+  return result;
+};
+
+const deleteReview = async (movieId, checkReview) => {
+  await dbDataSource.query(
+    `
+    DELETE FROM movie_reviews
+    WHERE movie_reviews.id = ?
+    `,
+    [checkReview]
+  );
+  let totalReview = await dbDataSource.query(
+    `
+    SELECT COUNT(id) as id
+    FROM movie_reviews
+    WHERE movies_id = ?
+    `,
+    [movieId]
+  );
+  totalReview = totalReview[0].id;
+
+  if (totalReview == 0) {
+    await dbDataSource.query(
+      `
+      UPDATE movies
+      SET average_rating = 0
+      WHERE id = ?
+      `,
+      [movieId]
+    );
+  } else {
+    let averageRating = await dbDataSource.query(
+      `
+      SELECT SUM(rating) as average_rating
+      FROM movie_reviews
+      WHERE movies_id = ?
+      `,
+      [movieId]
+    );
+    averageRating = averageRating[0].average_rating;
+
+    await dbDataSource.query(
+      `
+      UPDATE movies
+      SET average_rating = ?
+      WHERE id = ?
+      `,
+      [averageRating / totalReview, movieId]
+    );
+  }
+};
+
+const checkUserMovie = async (userId, movieId) => {
+  const result = await dbDataSource.query(
+    `
+    SELECT * 
+    FROM orders as o
+    JOIN reservation_options as ro
+    ON o.reservation_options_id = ro.id
+    WHERE o.users_id = ?
+    AND ro.id = ?
+    `,
+    [userId, movieId]
+  );
+  return result;
+};
```

Comment:
```suggestion
const checkUserMovie = async (userId, movieId) => {
  const [result] = await dbDataSource.query(
    `
    SELECT EXISTS (
      SELECT o.id
      FROM orders as o
      JOIN reservation_options as ro
      ON o.reservation_options_id = ro.id
      WHERE o.users_id = ?
      AND ro.id = ?
    ) as exists
    `,
    [userId, movieId]
  );
  return !!parseInt(result.exists);
};
```
단순 유무 여부만 판단할 것이라면 SELECT EXISTS를 활용하고 값이 있는지 없는지를 함수의 Return 에서 !!(double bang)을 활용하여 True / false로 반환될 수 있도록 수정해주세요. 성능 및 비용절감에서 차이가 있습니다.